### PR TITLE
Add custom output path in show-dependencies

### PR DESCRIPTION
**Pull request by ismaelvb**
Add support for custom output path for the show-dependencies command

### Motivation:

The command show-dependencies by default loads the missing packages and prints to the stdout the logging of download activities. The default output for the show-dependencies command is also stdout. This does not allow on automatic processing of the output, to separate the dependencies output from the log of loading the packages.

### Modifications:

Added a new option to the show-dependencies command that allows to redirect the dependencies output to a separate file.
The changes only affect the show dependencies command setup, once the output stream of the dependencies was already customisable through an optional parameter.

### Result:

No existing behaviour will be affected. There will be a new optional argument (-o | --output-path) on the show-dependencies command which will allow to send the dependencies graph output to a local file.

### CODE REVIEW

**Comment by tomerd** [comment]
thanks for the PR @ismaelvb stdout can be fairly easily redirected to a file with `swift package show-dependencies > <path>` for example. is that not good enough for the use case you have in mind?

File: Sources/Commands/SwiftPackageTool.swift [hunk 1]
```diff
@@ -466,9 +466,20 @@ extension SwiftPackageTool {
...
         var format: ShowDependenciesMode = .text
 
+        @Option(name: [.long, .customShort("o") ], 
+                help: "The absolute or relative path to output the resolved dependency graph.")
+        var outputPath: AbsolutePath?
+
         func run(_ swiftTool: SwiftTool) throws {
             let graph = try swiftTool.loadPackageGraph()
-            dumpDependenciesOf(rootPackage: graph.rootPackages[0], mode: format)
+            let stream : OutputByteStream
```
Comment (tomerd) [inline Sources/Commands/SwiftPackageTool.swift:475]: let stream = outputPath.map { try LocalFileOutputByteStream($0) } ?? TSCBasic.stdoutStream.stream

**Comment by ismaelvb** [comment]
> thanks for the PR @ismaelvb stdout can be fairly easily redirected to a file with `swift package show-dependencies > <path>` for example. is that not good enough for the use case you have in mind?

It is enough when the dependencies are already downloaded, but on the first attempt like when it is done right after the clone,  calling the to show-dependencies prints also log entries from the fetching to stdout which then are also redirected.

Example below of current latest on main:

> % swift package show-dependencies 
> Fetching https://github.com/apple/swift-llbuild.git from cache
> Fetching https://github.com/apple/swift-tools-support-core.git from cache
> Fetching https://github.com/apple/swift-driver.git from cache
> Fetching https://github.com/apple/swift-argument-parser.git from cache
> Fetching https://github.com/apple/swift-crypto.git from cache
> Fetching https://github.com/jpsim/Yams.git from cache
> Cloning https://github.com/apple/swift-crypto.git
> Resolving https://github.com/apple/swift-crypto.git at 1.1.6
> Cloning https://github.com/apple/swift-tools-support-core.git
> Resolving https://github.com/apple/swift-tools-support-core.git at main
> Cloning https://github.com/apple/swift-argument-parser.git
> Resolving https://github.com/apple/swift-argument-parser.git at 0.4.3
> Cloning https://github.com/jpsim/Yams.git
> Resolving https://github.com/jpsim/Yams.git at 4.0.6
> Cloning https://github.com/apple/swift-llbuild.git
> Resolving https://github.com/apple/swift-llbuild.git at main
> Cloning https://github.com/apple/swift-driver.git
> Resolving https://github.com/apple/swift-driver.git at main
> .
>├── llbuild<https://github.com/apple/swift-llbuild.git@unspecified>
>├── swift-tools-support-core<https://github.com/apple/swift-tools-support-core.git@unspecified>
>├── swift-argument-parser<https://github.com/apple/swift-argument-parser.git@0.4.3>
>├── swift-driver<https://github.com/apple/swift-driver.git@unspecified>
> │   ├── llbuild<https://github.com/apple/swift-llbuild.git@unspecified>
> │   ├── swift-tools-support-core<https://github.com/apple/swift-tools-support-core.git@unspecified>
> │   ├── Yams<https://github.com/jpsim/Yams.git@4.0.6>
> │   └── swift-argument-parser<https://github.com/apple/swift-argum

Where the graph load fetch logs are also redirected to the log.

File: Sources/Commands/SwiftPackageTool.swift (hunk 1 above)
Comment (ismaelvb) [reply Sources/Commands/SwiftPackageTool.swift]: Updated, two _try_ were required, once the exception was being thrown by the map

**Comment by tomerd** [comment]
I see. @neonichu @abertelrud wdyt?

**Comment by tomerd** [comment]
@abertelrud @neonichu @elsh wdyt?

**Comment by abertelrud** [comment]
This does make sense to me, especially when combined with the option to dump .json files or GraphViz .dot files, which are most useful when emitted to separate files.

**Comment by abertelrud** [comment]
It might be good to add a unit test here to check that the right file gets created and stdout doesn't contain the output that now goes into the file, just to make sure this doesn't break in the future.

**Review by elsh — APPROVED** [review]
Lgtm. If you could add unit tests as @abertelrud mentioned that'd be great.

**Comment by tomerd** [comment]
+1 to add a test, see "testShowDependencies" for example you can use to "seed" such test

**Comment by ismaelvb** [comment]
Sure, I'll add them.

**Comment by neonichu** [comment]
Thanks for the PR, @ismaelvb.

I think, separately, we should also think about making sure that informational output and logging generally goes to stderr and not to stdout, as making redirects work properly is the main reason for why stderr exists at all.

**Comment by ismaelvb** [comment]
@abertelrud, @eish: Unit test added.

I added the json context to the UT and also checked the correct output with one dependent package.
Used "seed" context from `testSymlinkedDependency`.

@neonichu, I agree, although the `--output-path` option is also useful in some contexts.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@ismaelvb Yep, agreed. I think both are useful, the option and also the pursuit to make redirection of stdout work as expected.

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
thanks @ismaelvb

**Pull request merged**